pkg/kubelet/lifecycle/handlers.go

```diff
@@ -230,3 +232,131 @@ func noNewPrivsRequired(pod *v1.Pod) bool {
 	}
 	return false
 }
+
+func NewCapabilityAdmitHandler() PodAdmitHandler {
+	return &capabilityAdmitHandler{}
+}
+
+type capabilityAdmitHandler struct{}
+
+// Check whether we have the capabilities to run the specified pod.
+func (a *capabilityAdmitHandler) Admit(attrs *PodAdmitAttributes) PodAdmitResult {
+	pod := attrs.Pod
+	if !capabilities.Get().AllowPrivileged {
+		for _, container := range pod.Spec.Containers {
+			if securitycontext.HasPrivilegedRequest(&container) {
+				return PodAdmitResult{
+					Admit:   false,
+					Reason:  "Forbidden",
+					Message: fmt.Sprintf("pod with UID %q specified privileged container, but is disallowed", pod.UID),
+				}
+			}
+		}
+		for _, container := range pod.Spec.InitContainers {
+			if securitycontext.HasPrivilegedRequest(&container) {
+				return PodAdmitResult{
+					Admit:   false,
+					Reason:  "Forbidden",
+					Message: fmt.Sprintf("pod with UID %q specified privileged init container, but is disallowed", pod.UID),
+				}
+			}
+		}
+	}
+
+	if pod.Spec.HostNetwork {
+		allowed, err := allowHostNetwork(pod)
+		if err != nil {
+			return PodAdmitResult{
+				Admit:   false,
+				Reason:  "Forbidden",
+				Message: err.Error(),
+			}
+		}
+		if !allowed {
+			return PodAdmitResult{
+				Admit:   false,
+				Reason:  "Forbidden",
+				Message: fmt.Sprintf("pod with UID %q specified host networking, but is disallowed", pod.UID),
+			}
+		}
+	}
+
+	if pod.Spec.HostPID {
+		allowed, err := allowHostPID(pod)
+		if err != nil {
+			return PodAdmitResult{
+				Admit:   false,
+				Reason:  "Forbidden",
+				Message: err.Error(),
+			}
+		}
+		if !allowed {
+			return PodAdmitResult{
+				Admit:   false,
+				Reason:  "Forbidden",
+				Message: fmt.Sprintf("pod with UID %q specified host PID, but is disallowed", pod.UID),
+			}
+		}
+	}
+
+	if pod.Spec.HostIPC {
+		allowed, err := allowHostIPC(pod)
+		if err != nil {
+			return PodAdmitResult{
+				Admit:   false,
+				Reason:  "Forbidden",
+				Message: err.Error(),
+			}
+		}
+		if !allowed {
+			return PodAdmitResult{
+				Admit:   false,
+				Reason:  "Forbidden",
+				Message: fmt.Sprintf("pod with UID %q specified host ipc, but is disallowed", pod.UID),
+			}
+		}
+	}
+	return PodAdmitResult{Admit: true}
+}
+
+// allowHostNetwork determines whether the specified pod is allowed to use host networking
+func allowHostNetwork(pod *v1.Pod) (bool, error) {
+	podSource, err := kubetypes.GetPodSource(pod)
```

Comment:
why not just return only `error`?

Since `if !allowed`, you still need to return an error message, like `fmt.Sprintf("pod with UID %q specified host ipc, but is disallowed", pod.UID)`.

And this will simplify the code in method `Admit()` as well.